### ISSUE TITLE
Change class name or namespace for a type used in Reliable Collection state crash application

### ISSUE THREAD

**Issue by rasmus-s** [Azure/service-fabric-issues#1344]
Normal with data contracts you can change the class name and/or the namespace of the class as long you have the original Name and Namespace in the DataContract attribute. But this its not working when its used in Reliable Collection state.  

I have this class (Class1) that it used in a Reliable Dictionary (deployed and Num get updated in the RunAsync every sec), and I want to change the name from Class1 to let’s say Class2. I only change the class name not the DataContract attribute – when I deploy it I immediately got an exception/crash in C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code\Microsoft.ServiceFabric.Data.Impl.dll version: 6.03.176.9494, and process exist. No stacktrace or exception message.  

How to I change name and namespace?

 ```
   [DataContract(Name = "Class1", Namespace = "myNamespace1")]
    public class Class1
    {
        [DataMember]
        public string Key { get; set; }

        [DataMember]
        public int Num { get; set; }
    }
```

**Comment by masnider** [comment]
Hi @rasmus-s - any way you can grab the crash info? For example, this should be reproable locally, and/or the information is probably in windows event viewer. My bet is that this is a serialization exception. I expect that this is because this change is causing the serialization to be different between primaries and secondaries. Can you try the process outlined here and see if it makes a difference?

https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-work-with-reliable-collections#schema-versioning-upgrades

**Comment by rasmus-s** [comment]
@masnider  Yes, I can grab the exception from the event viewer. See a the bottom. 
I dont change the DataContract attributes so all old partitions as new partitions should be able to deserialize the class, but its not the case. 
Maybe im wrong but should it not be okay to change the class name in c# code as long the it have the DataContracts attributes? or is SF not honor the DataContract and DataMember attributes ?

How to reproduce:
1. Create a new Service fabric Application with a stateful service (targetReplica=3).
2. Change the RunAsync to look like below.
3. Deploy it on a cluster with a least 3 nodes in Azure and see its work
4. Change the Class name and redeploy
5. Then it throw the exception on first UD.

```
protected override async Task RunAsync(CancellationToken cancellationToken)
        {
            var myDictionary1 = StateManager.GetOrAddAsync<IReliableDictionary2<string, Class1>>("TestDicV1").Result;

            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();
                using (var tx = this.StateManager.CreateTransaction())
                {
                    var result = await myDictionary1.GetOrAddAsync(tx, "Loop", s => new Class1() { });
                    result.Num = result.Num + 1;

                    await myDictionary1.SetAsync(tx, "Loop", result);
                    await tx.CommitAsync();
                }

                await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
            }
        }

  [DataContract(Name = "Class1", Namespace = "myNamespace1")]
    public class Class1
    {
        [DataMember]
        public string Key { get; set; }

        [DataMember]
        public int Num { get; set; }
    }
```

Stacktrace on the Azure Node - when deploy the changed class name:
```
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: b06d8a25-7a18-4327-b9c3-a3006a64a86a@131867818804222853@fabric:/StateManager: Below type used in Reliable Collection urn:TestDicV3/dataStore could not be loaded. This commonly indicates that the user application is not backwards/forwards compatible. Common compatibility bugs that lead to this error are adding a new type or changing an assembly name without two phase upgrade, or removing a type. If this was caused by user's backwards/forwards compatibility bug, one way to mitigate the issue is to force the upgrade through without safety checks.\r\n\r\nSystem.Fabric.Store.TStore`5[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Stateful1.Class3, Stateful1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[Microsoft.ServiceFabric.Data.Collections.DefaultComparer`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.ServiceFabric.Data.Collections.DefaultEqualityComparer`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.ServiceFabric.Data.Collections.EmptyRangePartitioner`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
```

**Comment by zuhairp** [comment]
That's expected. Reliable Collections persist the key and value type name of collections that are created, so changing the type of an existing collection is not supported. You can workaround this with a multi-phase deployment that creates a new collection with the existing collection's data:

**Phase 1**: Deploy a new version that has both ```Class1``` and ```Class2```. Don't add any code that actually uses ```Class2```
**Phase 2**: Deploy a new version that adds data to "TestDicV1" (which uses ```Class1```) and "TestDicV2" (which uses ```Class2```). You can do this by attaching to rebuild and change [notifications](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-notifications#reliable-dictionary-notifications) for TestDicV1. In the notification handlers, update TestDicV2.
**Phase 3**: Deploy a new version that switches to using TestDicV2 (no longer need notifications) and removes TestDicV1 from StateManager.

Wait for log truncation - by default, this happens every 50MB of writes (adds, updates, and removes) to reliable collections. We are hoping to add tools to explicitly force truncation in our next release.

**Phase 4**: Deploy a new version that removes ```Class1```.

**Comment by rasmus-s** [comment]
@zuhairp thanks for the detailed guide - i will try it out tomorrow at work. But this mean that both the name and namespace properties in the DataContract attributes cannot be used to change the name or namespace of a class when working with Service Fabric - like you can when you work with DataContracts as eg. wire protocol.

For the Phase 4 (remove Class1) - I had created another github issue: (https://github.com/Azure/service-fabric-issues/issues/1352 - Failed to remove classes that had been used earlier in a removed Reliable Collection) due when I try to remove the class after first remove the Reliable Collection state from the StateManager in a earlier phase, its throw a exception. Can you please assist on that?

**Comment by zuhairp** [comment]
Hey, I just realized that you will be unable to do the copy in the rebuild notification I mentioned in **Phase 2** because the partition does not have write status at that point. Phase 2 should be as follows:

**Phase 2**: Deploy a new version where all writes (adds and updates) go to "TestDicV2" (which uses ```Class2```). Reads first check "TestDicV1" then "TestDicV2". Concurrently, enumerate "TestDicV1" and copy to "TestDicV2" with ``TryAdd`` API.

That approach gets complicated if you are expecting Removes as well, since it might get Added and Removed from TestDicV2 then Added again during the copy of TestDicV1. 

If a period without write availability is okay, it's much easier to block all incoming writes while you copy the dictionaries.

**Comment by rasmus-s** [comment]
I have testet the 4 phase deploy to change the CLR class name of the root type that its use in the Reliable Collection - and its seems to work except of Phase 4. Properly due to, I dont know when it have hit the 50mb log, and that way not know if there has been a log truncation or not. Are there a event that I can listen on? Or ask for last log truncation has happen?

To notice, I can changing CLR class name or namespace of **nested** classes to the Reliable Collection root data class, by using the DataCrontract attributes. So here i dont need to follow the 4 phase deploy.

Its still seems to be a lot of work to change CLR name or namespace of root classes used for the Reliable Collections.  
My understanding is that SF try to find the CLR classes that its use as root data class in the Reliable Collection, and if its not found the class it will throw the deserialization exception. Could SF be extended to also check if there is a CLR class with a DataContracts that match the namespace and name - and this way make its simpler for a developer to change names and namespaces in your project.
Alternative add a way to inform the stateManager that it should use class X instead for deserialization.

**Comment by zuhairp** [comment]
There isn't a straightforward way to check whether a log truncation has happened. You might be able to tell by the file times in the application's work folder. You can look at the file time for StateManager.cpt. If it's after you deployed Phase 3, then that means a checkpoint has occured, and you can continue with Phase 4. 
Also, thanks for the feedback. We'll take those ideas into consideration, especially since we're actively working on improving the experience around this scenario.

**Comment by zuhairp** [comment]
I actually meant to say 6.5, sorry for the confusion. 6.4 had already been "released" from the product group's perspective, so next release meant 6.5 to me when I wrote that.

**Comment by rasmus-s** [comment]
oh, I had hope for something soon. Do you have a rough estimate when 6.5 will be released - januar/feb ?

**Comment by zuhairp** [comment]
I'm not really sure when it will be released - it will be at least a few months

**Comment by JohnNilsson** [comment]
Just want to add my +1. Was just hit by this after trying to move a class to another assembly. Keeping the same namespace and class name I wasn't really expecting to discover that the serializer no longer could read the data due to the assembly change.
